lib/evaluate/constant.cc

```diff
@@ -46,6 +47,64 @@ bool IncrementSubscripts(ConstantSubscripts &indices,
   return false;  // all done
 }
 
+bool IncrementSubscripts(ConstantSubscripts &indices,
+    const ConstantSubscripts &shape, const ConstantSubscripts &lbound,
+    const std::vector<int> &dimOrder) {
+  int rank{GetRank(shape)};
+  CHECK(GetRank(indices) == rank);
+  CHECK(GetRank(lbound) == rank);
+  CHECK(static_cast<int>(dimOrder.size()) == rank);
+  for (int j{0}; j < rank; ++j) {
+    auto lb{lbound[j]};
+    ConstantSubscript k{dimOrder[j]};
+    CHECK(indices[k] >= lb);
+    if (++indices[k] < lb + shape[k]) {
+      return true;
+    } else {
+      CHECK(indices[k] == shape[k] + lb);
+      indices[k] = lb;
+    }
+  }
+  return false;  // all done
+}
+
+std::optional<std::vector<int>> IsValidDimensionOrder(
+    int rank, const std::optional<std::vector<int>> &order) {
+  if (!order.has_value()) {
+    std::vector<int> result;
+    for (int j{0}; j < rank; ++j) {
+      result.push_back(j);
+    }
+    return {std::move(result)};
```

Comment:
Done. I was a bit paranoid with this after #527 and #532 and was not willing to take the risk to break Ubuntu 18.04.... After your comment I checked on Ubuntu and you are correct, this is not an issue here. I was curious and could not take leaving in fear anymore, so I investigated a bit more, the bug #532 seems to be a g++ compiler bug (not an stdlib bug). I could reproduce with simple classes and come with the following workaround rule:
 
> One need to explicitly move an lvalue result that should be automatically moved by the compiler when:
    *  it can only be move constructed into the result type (no copy constructor), and
    *  where the result type constructor is defined using perfect forwarding (template<typename A> Constructor(A&& x): x_{std::forward(x)})

`result` is copyable, so it will be copy constructed into an `std::optional<std::vector<int>>` with Ubuntu 18.04 g++ (which is technically wrong, since it should have selected the move constructor, but at least, it does not crash).

Reply:
Can you please add a note about this bug with copy elision in older GCC compilers to the documentation somewhere so that we don't forget it?  Thank you.

Reply:
Sure I added a note at the end of the C++ style document in a new section about workaround for C++ compiler bugs.